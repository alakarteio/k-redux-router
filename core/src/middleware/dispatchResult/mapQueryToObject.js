/* eslint-env browser */

// https://stackoverflow.com/questions/8648892/convert-url-parameters-to-a-javascript-object
export default () => {
  const { location } = window

  if (location.search.length < 2) return {}

  const search = location.search.substring(1)
  return JSON.parse(`{"${decodeURI(search).replace(/"/g, '\\"').replace(/&/g, '","').replace(/=/g, '":"')}"}`)
}
