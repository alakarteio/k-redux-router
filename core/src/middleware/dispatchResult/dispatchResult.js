import routeFound from '../routeFound'
import mapQueryToObject from './mapQueryToObject'

export default (options, reducer) => (store) => {
  const innerWindow = options.window || window

  if (innerWindow && innerWindow.location && innerWindow.location.pathname) {
    // find route (& path params)
    let pathParams
    const route = reducer.getState(store.getState()).routes.array.find((r) => {
      pathParams = r.href.regexp.exec(innerWindow.location.pathname)
      return !!pathParams
    })

    // TODO: handle not found
    if (!route) return

    // attach names to path params
    pathParams = route.href.parsed
      .map(part => part.name)
      .filter(Boolean)
      .reduce(
        (acc, curr, index) => Object.assign({ [curr]: pathParams[index + 1] }, acc),
        {},
      )

    store.dispatch(routeFound({
      route,
      params: {
        path: pathParams,
        query: mapQueryToObject(innerWindow),
      },
    }))
  }
}
