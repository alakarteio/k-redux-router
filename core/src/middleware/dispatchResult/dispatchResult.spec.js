/* eslint-env jest */
import dispatchResult from './dispatchResult'

const match = res => () => expect(res).toMatchSnapshot()

describe('dispatchResult', () => {
  const reducer = {
    getState: state => state.router,
  }

  const store = {
    getState: () => ({
      router: {
        routes: {
          array: [
            {
              href: {
                regexp: /\/some\/(.*)\/name/,
                parsed: [{ name: 'id' }]
              },
            },
          ],
        },
      },
    }),
    dispatch: jest.fn(),
  }

  const options = {
    window: {
      location: {
        pathname: '/some/102/name',
        search: '?query=param',
      },
    },
  }

  beforeEach(() => {
    store.dispatch.mockClear()
  })

  it('should dispatch nothing when no location is given', () => {
    dispatchResult({ ...options, window: {} }, reducer)(store)

    expect(store.dispatch.mock.calls.length).toBe(0)
  })

  it('should dispatch nothing when no pathname is given', () => {
    dispatchResult({ ...options, window: { location: {} } }, reducer)(store)

    expect(store.dispatch.mock.calls.length).toBe(0)
  })

  it('should dispatch nothing if no route is found', () => {
    dispatchResult({ ...options, window: { location: { pathname: 'not-found' } } }, reducer)(store)

    expect(store.dispatch.mock.calls.length).toBe(0)
  })

  it('should dispatch ROUTE_FOUND action with path and query params', () => {
    dispatchResult(options, reducer)(store)

    const { calls } = store.dispatch.mock
    expect(calls.length).toBe(1)
    match(calls[0])()
  })
})
