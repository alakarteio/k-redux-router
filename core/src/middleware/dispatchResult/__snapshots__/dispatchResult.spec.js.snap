// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dispatchResult should dispatch ROUTE_FOUND action with path and query params 1`] = `
Array [
  Object {
    "payload": Object {
      "found": true,
      "params": Object {
        "path": Object {
          "id": "102",
        },
        "query": Object {
          "query": "param",
        },
      },
      "route": Object {
        "href": Object {
          "parsed": Array [
            Object {
              "name": "id",
            },
          ],
          "regexp": /\\\\/some\\\\/\\(\\.\\*\\)\\\\/name/,
        },
      },
    },
    "type": "@@router/ROUTE_FOUND",
  },
]
`;
